packages/snaps-controllers/src/snaps/location/npm.ts

```diff
@@ -0,0 +1,353 @@
+      pump(
+        getNodeStream(tarballResponse),
+        // The "gz" in "tgz" stands for "gzip". The tarball needs to be decompressed
+        // before we can actually grab any files from it.
+        createGunzipStream(),
+        createTarballStream(
+          `${canonicalBase}/${this.meta.packageName}/`,
+          this.files,
+        ),
+        (error) => {
+          error ? reject(error) : resolve();
+        },
+      );
+    });
+  }
+}
+
+/**
+ * Fetches the tarball (`.tgz` file) of the specified package and version from
+ * the public npm registry. Throws an error if fetching fails.
+ *
+ * @param packageName - The name of the package whose tarball to fetch.
+ * @param versionRange - The SemVer range of the package to fetch. The highest
+ * version satisfying the range will be fetched.
+ * @param registryUrl - The URL of the npm registry to fetch the tarball from.
+ * @param fetchFunction - The fetch function to use. Defaults to the global
+ * {@link fetch}. Useful for Node.js compatibility.
+ * @returns A tuple of the {@link Response} for the package tarball and the
+ * actual version of the package.
+ */
+async function fetchNpmTarball(
+  packageName: string,
+  versionRange: SemVerRange,
+  registryUrl: URL | string,
+  fetchFunction: typeof fetch,
+): Promise<[ReadableStream, SemVerVersion]> {
+  const packageMetadata = await (
+    await fetchFunction(new URL(packageName, registryUrl).toString())
+  ).json();
+
+  if (!isObject(packageMetadata)) {
+    throw new Error(
+      `Failed to fetch package "${packageName}" metadata from npm.`,
+    );
+  }
+
+  const versions = Object.keys((packageMetadata as any)?.versions ?? {}).map(
+    (version) => {
+      assertIsSemVerVersion(version);
+      return version;
+    },
+  );
+
+  const targetVersion = getTargetVersion(versions, versionRange);
+
+  if (targetVersion === null) {
+    throw new Error(
+      `Failed to find a matching version in npm metadata for package "${packageName}" and requested semver range "${versionRange}"`,
+    );
+  }
+
+  const tarballUrlString = (packageMetadata as any).versions?.[targetVersion]
+    ?.dist?.tarball;
+
+  if (!isValidUrl(tarballUrlString) || !tarballUrlString.endsWith('.tgz')) {
+    throw new Error(
+      `Failed to find valid tarball URL in npm metadata for package "${packageName}".`,
+    );
+  }
+
+  // Override the tarball hostname/protocol with registryUrl hostname/protocol
+  const newRegistryUrl = new URL(registryUrl);
+  const newTarballUrl = new URL(tarballUrlString);
+  newTarballUrl.hostname = newRegistryUrl.hostname;
+  newTarballUrl.protocol = newRegistryUrl.protocol;
+
+  // Perform a raw fetch because we want the Response object itself.
+  const tarballResponse = await fetchFunction(newTarballUrl.toString());
+  if (!tarballResponse.ok) {
+    throw new Error(`Failed to fetch tarball for package "${packageName}".`);
+  }
+  const stream = await tarballResponse.blob().then((blob) => blob.stream());
```

Comment:
I believe this can be simplified now that we arent using `cross-fetch` anymore: https://github.com/MetaMask/snaps-monorepo/pull/1013

Reply:
Nice